tests: line: Use fixed-point arithmetic

maromaSamsa opened the pull request:
Instructions: https://hackmd.io/@maromaSamsa/HkjefPbFs

Optimize the performance by replacing floating arithmetic with fixed-point arithmetic counterpart.

jserv commented on tests/line.c:
```diff
@@ -150,70 +150,274 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
+
+/* minimum value of Q format */
+#define QFMT_MIN 0x80000000
+
```
Remove one blank line.

jserv commented on tests/line.c:
```diff
@@ -150,70 +150,274 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
-#include <math.h>  // ceilf(), floorf(), fminf(), fmaxf(), sinf(), cosf(), sqrtf()
 #include <string.h>  // memset()
+#define Q (20)
+#define PI_Q (1686629713 >> (29 - Q))
```
Rename to `Q_PI` for consistency.

jserv commented on tests/line.c:
```diff
@@ -150,70 +150,274 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
+#define QFMT_MIN 0x80000000
+
+
+/* addition of Q format value*/
```
Should be `/* addition of Q format value */`

jserv commented on tests/line.c:
```diff
@@ -150,70 +150,274 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
+        q_fmt t = j * q_div(PI_Q, I2Q(64));
+        q_fmt r = (j + 1) * (1 << (Q - 1));
+        for (int i = 1; i <= 64;
+             i++, t = q_add(t, q_mul(I2Q(2), q_div(PI_Q, I2Q(64))))) {
```
Move the statement `t = q_add(t, q_mul(I2Q(2), q_div(PI_Q, I2Q(64))))` right after `lineSDFAABB`.

jserv commented on tests/line.c:
```diff
@@ -150,70 +150,274 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
-    float cx = W * 0.5f, cy = H * 0.5f;
+    q_fmt cx = W * (1 << (Q - 1)), cy = H * (1 << (Q - 1));
+    q_fmt ct, st;
     for (int j = 0; j < 5; j++) {
```
You can add some comments.

maromaSamsa commented:
Hi @jserv, I have formatted the code in clang-format-12 and done the  changes you proposed above. Please check.

Thank you

jserv commented on tests/line.c:
```diff
@@ -150,70 +150,277 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
+}
+
+/* return the largest integral value that is not greater than x */
+static inline q_fmt floorq(q_fmt x)
```
Rename to `q_floor`. Change the corresponding functions, ensuring prefix `q_`.

maromaSamsa replied:
Should I also rename basic arithmetic functions?  e.g. q_add() rename to qadd().

jserv replied:
No, let's unify with the consistent `q_` prefix.

jserv commented on tests/line.c:
```diff
@@ -150,70 +150,277 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
+    return (offset >= 0) ? res >> offset : res << (-offset);
+}
+
+static inline q_fmt cos_half(q_fmt cosx)
```
Can you explain `cos_half` and `sin_half`?

maromaSamsa replied:
Certainly, I will add an annotation above.

jserv replied:
Unify the naming scheme of helper functions. i.e., prefix `q_`.

jserv reviewed (changes requested):
Use `git rebase -i` to squash and read the article [How to Write a Git Commit Message
](https://cbea.ms/git-commit/) carefully.

You shall mention the benefits (with measurements) of the proposed changes.

jserv commented on tests/line.c:
```diff
@@ -150,70 +150,277 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
+    }
+
+    if (region == 0b10 || region == 0b01)
+        *cos_t *= -1;
```
Can you use bitwise operations to toggle the signedness?

jserv commented on tests/line.c:
```diff
@@ -150,70 +150,277 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
+
+    /* left shift x as more as possible */
+    int offset = 0;
+    for (q_fmt i = x; !(0x40000000 & i); i <<= 1) {
```
Rewrite as following:
```c
for (q_fmt i = x; !(0x40000000 & i); i <<= 1)
        ++offset;
```

jserv commented on tests/line.c:
```diff
@@ -150,70 +150,277 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
+    x <<= offset;
+
+    /* shift bit to the highest bit 1' in x */
+    while (bit > x) {
```
Ditto. Shorten the code.

jserv replied:
?

jserv commented on tests/line.c:
```diff
@@ -150,70 +150,277 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
+}
+
+/* return the nonnegative square root of x */
+static inline q_fmt sqrtq(q_fmt x)
```
Can you check the following materials accordingly?
* [Fixed Point Square Root](https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=71801aa714a149272e58640f2586ddb3a4fa397b)
* [Fixed-Point Arithmetic](https://vanhunteradams.com/FixedPoint/FixedPoint.html)
* [Rootfinding - Fixed Point Method](https://youtu.be/PzHejhnM5sM)

I would expect the high-level descriptions for square root implementation.

maromaSamsa replied:
I have read the following article, it seems that the fastest algorithm which the article recommended is converting the fixed point value to the floating point value to do 'sqrtf()', and then converting back to the fixed point value result.

Should I change the algorithm into this way? Or just use the original one I made (converging bit by bit)

jserv replied:
Always measure if something is in doubt.

maromaSamsa replied:
Yes, I have already measured the performance between them in my previous term project:

For performing each algorithm 1000 times and compare their clock cycle:

Convert to floating point and call sqrtf(): 228442
My algorithm: 462650

The problem I concern is that our mission is getting rid of floating point arithmetic, this is not a good idea of still using them.

jserv commented on tests/line.c:
```diff
@@ -150,70 +150,277 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
+#define Q (20)
+#define Q_PI (1686629713 >> (29 - Q))
+
+/* 32-bit Q format */
```
Change the comment to
> 32-bit Q notation to specify a fixed point number format.

jserv commented on tests/line.c:
```diff
@@ -150,70 +150,277 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
+/* 32-bit Q format */
+typedef int32_t q_fmt;
+
+/* 64-bit Q format buffer */
```
You shall add more comments to explain why `q_buf` exists.

maromaSamsa commented:
Except for the high-level descriptions for square root implementation, several annotations have been added. For bit-wise operation of signedness, `x = ~x+1` can replace x *= -1.

I will squash those commit to one if all adjustments are finished.

jserv commented on tests/line.c:
```diff
@@ -150,70 +150,288 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
+}
+
+/* get both sin and cos value in the same radius */
+static inline void q_sin_and_cos(q_fmt radius, q_fmt *sin_t, q_fmt *cos_t)
```
Simply follow the naming rules as [sincos](https://man7.org/linux/man-pages/man3/sincos.3.html). i.e., `q_sincos`.

jserv commented on tests/line.c:
```diff
@@ -150,70 +150,288 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
+#define Q_PI (1686629713 >> (29 - Q))
+
+/* 32-bit Q notation to specify a fixed point number format. */
+typedef int32_t q_fmt;
```
If you read [CONTRIBUTING.md](https://github.com/sysprog21/rv32emu/blob/master/CONTRIBUTING.md) carefully, you would be aware of suffix `_t` used in typedef'd types. Here, for the sake of consistency, we can rewrite as following:
```c
typedef int32_t qfixed_t;
typedef int64_t qbuf_t;
```

jserv commented on tests/line.c:
```diff
@@ -150,70 +150,288 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
+}
+
+/* half-width formula for cos(x) */
+static inline q_fmt q_cos_half(q_fmt cosx)
```
Since both `q_cos_half` and `q_sin_half` are invoked once, it is not worthy to declare a short inline function. You can simply move the operations into `q_sincos`.

jserv commented on tests/line.c:
```diff
@@ -150,70 +150,281 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
-    p[0] = (uint8_t) (p[0] * (1 - alpha) + r * alpha * 255);
-    p[1] = (uint8_t) (p[1] * (1 - alpha) + g * alpha * 255);
-    p[2] = (uint8_t) (p[2] * (1 - alpha) + b * alpha * 255);
+    p[0] = (uint8_t) Q2I(
```
Rewrite as following:
```c
#define PUT(v) \
    p[v] = (uint8_t) Q2I(q_add(p[v] * q_add((1 << Q), -alpha), q_mul(g, alpha) * 255))
    PUT(0);
    PUT(1);
    PUT(2);
#undef PUT
```

It is a trick to write more maintainable code.

jserv reviewed (changes requested):
Improve the git commit message. You shall express the motivation and benefits of the proposed changes. Show the numbers.

jserv commented:
Thank @maromaSamsa for contributing!

maromaSamsa commented:
Thank @jserv let me have an opportunity contributing to this great work.

maromaSamsa committed:
Using fixed-point arithmetic

Optimize the performance by replacing floating arithmetic with
fixed-point arithmetic counterpart, which allows the embedded
microprocessors that didn't have floating-point unit (FPU) doing
mechanical calculation in high speed and/or low power consumption way.

This change will reduce program execution time by 14%.

Reference: https://hackmd.io/@maromaSamsa/HkjefPbFs
